lib/gssapi/sanon/canonicalize_name.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2019-2020, AuriStor, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ *
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in
+ *   the documentation and/or other materials provided with the
+ *   distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "sanon_locl.h"
+
+OM_uint32 GSSAPI_CALLCONV
+_gss_sanon_canonicalize_name(OM_uint32 *minor,
+			     gss_const_name_t src_name,
+			     const gss_OID mech_type,
+			     gss_name_t *dest_name)
+{
+    *minor = 0;
+
+    if (src_name == GSS_C_NO_NAME)
+	return GSS_S_BAD_NAME;
+
+    if (_gss_sanon_is_anonymous_name_p(src_name))
+	*dest_name = _gss_sanon_wellknown_anon_name();
```

Comment:
Is there any reason not to just always do this?

An SAnon mech name (MN) should probably always have the same anon form.  That is, a `gss_name_t` obtained from `gss_accept_sec_context()`, `gss_inquire_cred*()`, `gss_inquire_context()`, or `gss_canonicalize_name()` should probably all have exactly the same display form and name-type.

Reply:
I might be missing something but, if you call `gss_canonicalize_name()` on a non-anonymous name, and then pass that to `gss_init_sec_context()`, you may get anonymous authentication when you didn't want it. See the [draft](https://datatracker.ietf.org/doc/html/draft-howard-gss-sanon#section-4.1).

Reply:
You also explained on IM.  I'll resolve this now then.